File: benchmarks/state-machines/cross/validate.go
// Go standard library: unicode/utf8.Valid over the shared input.
package main

import (
	"fmt"
	"os"
	"time"
	"unicode/utf8"
)

func main() {
	buf, err := os.ReadFile("input.bin")
	if err != nil {
		panic(err)
	}
	best := time.Duration(1 << 62)
	ok := false
	for r := 0; r < 5; r++ {
		t0 := time.Now()
		ok = utf8.Valid(buf)
		if d := time.Since(t0); d < best {
			best = d
		}
	}
	n := float64(len(buf))
	fmt.Printf("%-34s %6.2f ns/byte  %6.2f GB/s\n", "Go unicode/utf8.Valid", float64(best.Nanoseconds())/n, n/float64(best.Nanoseconds()))
	if !ok {
		os.Exit(1)
	}
}

File: benchmarks/state-machines/cross/validate.rs
// Rust standard library: std::str::from_utf8 over the shared input.
use std::time::Instant;
fn main() {
    let buf = std::fs::read("input.bin").expect("run gen_input first");
    let mut best = f64::MAX;
    let mut ok = false;
    for _ in 0..5 {
        let t0 = Instant::now();
        ok = std::str::from_utf8(&buf).is_ok();
        let t = t0.elapsed().as_nanos() as f64;
        if t < best { best = t; }
    }
    let n = buf.len() as f64;
    println!("{:<34} {:6.2} ns/byte  {:6.2} GB/s", "Rust std::str::from_utf8", best / n, n / best);
    if !ok { std::process::exit(1); }
}
